apollo-router/src/compute_task.rs

```diff
@@ -0,0 +1,48 @@
+use std::sync::OnceLock;
+
+use tokio::sync::oneshot;
+
+static POOL: OnceLock<threadpool::ThreadPool> = OnceLock::new();
+
+/// Drop-in replacement for Tokio’s [`spawn_blocking`]
+/// but intended for for tasks that keep a CPU core active: it uses a thread pool
+/// limited to the number of available CPU cores.
+///
+/// `spawn_blocking` on the other hand appears to be intended for tasks that pause a thread
+/// in a blocking syscall: `max_blocking_threads` is configurable but defaults to 512,
+/// which is too high for tasks where the CPU is actively running.
+/// Configuring it to $NUM_CPUS feels risky in case a library we use relies on `spawn_blocking`
+/// in the originally intended way.
+///
+/// [`spawn_blocking`]: https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html
+///
+/// ## Example
+///
+/// Doing this in an async function can work but isn't great for Tokio performance:
+///
+/// ```no_run
+/// let output = some_heavy_computation();
+/// ```
+///
+/// It can be replaced with:
+///
+/// ```no_run
+/// use crate::compute_task;
+///
+/// let output = compute_task::execute(|| some_heavy_computation()).await;
+/// ```
+pub(crate) fn execute<T, F>(f: F) -> oneshot::Receiver<T>
+where
+    F: FnOnce() -> T + Send + 'static,
+    T: Send + 'static,
+{
+    // https://docs.rs/threadpool/latest/threadpool/struct.Builder.html#method.num_threads
+    // > defaults the number of threads to the number of CPUs
+    let pool = POOL.get_or_init(|| threadpool::Builder::new().build());
```

Comment:
We don't need to / want to reserve some CPUs for Tokio tasks? Obviously we have no control over how threads get scheduled, I just wonder if we should pull back on the default number of threads since we know we have another task scheduler.

Reply:
Why don't we use tokio's `spawn_blocking` method? It may work better together with I/O tasks.

Reply:
Maybe! If we have more threads that want CPU time than available CPU cores things don’t break down immediately, the kernel scheduler preempts threads and tries to distribute time fairly. So at 1.5x or even 2x utilization (over compute threads + tokio threads) we may have some context switching overhead but it shouldn’t be as bad as with 512 threads. We can tweak the number but it’s hard to guess what a good value is: 50% of `num_cpus()`? 80%? We could make it configurable but most users may not know any more about what a good value for this obscure number is.

Reply:
Let's merge this as is because it's already an improvement on `spawn_blocking()` and we can think about the best way to slice up CPU resource once this change is in place.